services/serviceutils/src/main/java/tech/pegasys/teku/service/serviceutils/layout/SeparateServiceDataDirLayout.java

```diff
@@ -52,8 +52,9 @@ public void migrateIfNecessary() throws IOException {
   private void performMigration() throws IOException {
     STATUS_LOG.migratingDataDirectory(baseDir);
     final File dataSubdir = baseDir.resolve("data").toFile();
-    if (dataSubdir.isDirectory()) {
-      for (final File file : dataSubdir.listFiles()) {
+    File[] fileList = dataSubdir.listFiles();
+    if (fileList != null) {
+      for (final File file : fileList) {
         Files.move(file.toPath(), baseDir.resolve(file.getName()));
       }
```

Comment:
Thanks for submitting this, but I'm unclear what the practical aim of this is.

This directory is under Teku's data directory.  If we operate under the assumption that other processes are making unexpected changes in that directory we would need to account for the database being randomly corrupted, deleted or any number of other conditions which are impossible to handle in any way except crashing.

Even with this change, if the files in the directory were to change between listing them and moving them the operation would fail so this doesn't actually avoid the potential for race conditions.

Reply:
The practical aim is preventing the code from throwing an unhandled NPE.

Reply:
The same bug is reported [here](https://dzone.com/articles/folderlistfiles-may-produce-nullpointerexception).
It is safer to call listFiles once. 
This is an easy fix, probably takes less time than writing all the comments :-)

Unhandled NPE ranks high (13) in the [25 Most Dangerous Software Weaknesses](https://cwe.mitre.org/top25/archive/2020/2020_cwe_top25.html).
 [CWE 476](https://cwe.mitre.org/data/definitions/476.html) provides a more detailed description of this weakness.

Reply:
But the change doesn't avoid a crash if the files are changing under us.  And we *want* to crash if something else is modifying our data directory unexpectedly. The `NullPointerException` that would result wouldn't have any dangerous side effects - Teku would simply fail to start which is exactly the behaviour we want in this case.

So again I'd ask, what is the practical implication of this issue. What actual behaviour is exhibited that we need to change? Because as far as I can see this change moves the risk further down the code to where files are actually being moved which actually makes things worse.